### ISSUE TITLE
yarn 2 Qualified path resolution failed - none of the candidates can be found on the disk.

### ISSUE THREAD

**Issue by mdbetancourt** [loopbackio/loopback-datasource-juggler#1866]
![image](https://user-images.githubusercontent.com/22435967/93167921-af7c5880-f6ef-11ea-88e1-1dbd77a7c1c7.png)
looks like the problem is in the following code
```ts

function connectorModuleNames(name) {
  const names = []; // Check the name as is
  if (!name.match(/^\//)) {
    names.push('./connectors/' + name); // Check built-in connectors
    if (name.indexOf('loopback-connector-') !== 0) {
      names.push('loopback-connector-' + name); // Try loopback-connector-<name>
    }
  }
  // Only try the short name if the connector is not from StrongLoop
  if (['mongodb', 'oracle', 'mysql', 'postgresql', 'mssql', 'rest', 'soap', 'db2', 'cloudant']
    .indexOf(name) === -1) {
    names.push(name);
  }
  return names;
}
```
trying to check for rest file in connector folder but there no exists

i fixed it with

```ts
function connectorModuleNames(name) {
  const names = []; // Check the name as is
  if (!name.match(/^\//)) {
    if(['kv-memory', 'memory', 'transient'].indexOf(name) !== -1) {
      console.log(name)
      names.push('./connectors/' + name); // Check built-in connectors

    } else if (name.indexOf('loopback-connector-') !== 0) {
      names.push('loopback-connector-' + name); // Try loopback-connector-<name>
    }
  }
  // Only try the short name if the connector is not from StrongLoop
  if (['mongodb', 'oracle', 'mysql', 'postgresql', 'mssql', 'rest', 'soap', 'db2', 'cloudant']
    .indexOf(name) === -1) {
    names.push(name);
  }
  return names;
}
```

**Comment by mdbetancourt** [comment]
a workaround is use the full name 
{
connector: "loopback-connector-rest" instead of "rest"
}

**Comment by ferrantejake** [comment]
I am having the same issue with the mongodb connector. It is not resolved by using the connectors full name.

```shell
Cannot start the application. Error: Qualified path resolution failed - none of the candidates can be found on the disk.

Source path: C:\Users\jakef\code\examplerepo\.yarn\cache\loopback-datasource-juggler-npm-4.26.0-67d3cf0749-75cc866df5.zip\node_modules\loopback-datasource-juggler\lib\connectors\loopback-connector-mongodb
Rejected candidate: C:\Users\jakef\code\examplerepo\.yarn\cache\loopback-datasource-juggler-npm-4.26.0-67d3cf0749-75cc866df5.zip\node_modules\loopback-datasource-juggler\lib\connectors\loopback-connector-mongodb
Rejected candidate: C:\Users\jakef\code\examplerepo\.yarn\cache\loopback-datasource-juggler-npm-4.26.0-67d3cf0749-75cc866df5.zip\node_modules\loopback-datasource-juggler\lib\connectors\loopback-connector-mongodb.js
Rejected candidate: C:\Users\jakef\code\examplerepo\.yarn\cache\loopback-datasource-juggler-npm-4.26.0-67d3cf0749-75cc866df5.zip\node_modules\loopback-datasource-juggler\lib\connectors\loopback-connector-mongodb.json
Rejected candidate: C:\Users\jakef\code\examplerepo\.yarn\cache\loopback-datasource-juggler-npm-4.26.0-67d3cf0749-75cc866df5.zip\node_modules\loopback-datasource-juggler\lib\connectors\loopback-connector-mongodb.node
```

**Comment by mdbetancourt** [comment]
> I am having the same issue with the mongodb connector. It is not resolved by using the connectors full name.

do you have installed mongodb connector? try using .yarnrc.yml to specify the dependency

**Comment by ferrantejake** [comment]
Yes, I do. The issue lies in the way datasource-juggler tries to look up multiple versions of the connector by looking in different locations and fails entirely when it fails on any of the locations it tries to look up. See this pull request.
https://github.com/strongloop/loopback-datasource-juggler/pull/1883

**Comment by bajtos** [comment]
Another option is to pass the connector object instead of the module name.

```js
// LB4 datasource config
const config = {
  connector: require('loopback-connector-mongodb'),
  // ...
};
```

**Comment by instantlinux** [comment]
I've upgraded to yarn 2.4.1 and hit this irreconcilable error in CI; these problems are maddening and time-consuming for those of us not native to the javascript frontend world:
```
Qualified path resolution failed - none of the candidates can be found on the disk.
    Source path: /builds/instantlinux/apicrud-ui/.yarn/$$virtual/@blackbox-vision-ra-language-spanish-virtual-229dab0f27/0/cache/@blackbox-vision-ra-language-spanish-npm-3.12.2-f02804f106-3619e7ae0a.zip/node_modules/@blackbox-vision/ra-language-spanish/
    Rejected candidate: /builds/instantlinux/apicrud-ui/.yarn/$$virtual/@blackbox-vision-ra-language-spanish-virtual-229dab0f27/0/cache/@blackbox-vision-ra-language-spanish-npm-3.12.2-f02804f106-3619e7ae0a.zip/node_modules/@blackbox-vision/ra-language-spanish/
```
This comes from the following line in `App.js`:
```
import spanishMessages from '@blackbox-vision/ra-language-spanish';
```
Somehow it's getting confused by the `@` or something, generating an incorrect path for that one (out of dozens) of packages. My current workaround is to turn off the feature that this `import` enables: Spanish language translations. But that's hardly satisfying. It's hard to gather info to submit my own bug report, because the problem only happens in my CI builds (on throwaway containers).

**Comment by lensbart** [comment]
@instantlinux I’m running into a similar issue and think it’s related to Yarn 2 (PnP). In my case, the issue arises on AWS Elastic Beanstalk (Linux), I think with the first import that has an “@”.

Did you figure out a solution already? In any case, I’ll update here if/when I figure it out.

```
-------------------------------------
/var/log/nodejs/nodejs.log
-------------------------------------
    ^

Error: Qualified path resolution failed - none of the candidates can be found on the disk.

Source path: /var/app/current/.yarn/unplugged/@apollo-protobufjs-npm-1.0.5-a95d547915/node_modules/@apollo/protobufjs/src/index-minimal
Rejected candidate: /var/app/current/.yarn/unplugged/@apollo-protobufjs-npm-1.0.5-a95d547915/node_modules/@apollo/protobufjs/src/index-minimal
Rejected candidate: /var/app/current/.yarn/unplugged/@apollo-protobufjs-npm-1.0.5-a95d547915/node_modules/@apollo/protobufjs/src/index-minimal.js
Rejected candidate: /var/app/current/.yarn/unplugged/@apollo-protobufjs-npm-1.0.5-a95d547915/node_modules/@apollo/protobufjs/src/index-minimal.json
Rejected candidate: /var/app/current/.yarn/unplugged/@apollo-protobufjs-npm-1.0.5-a95d547915/node_modules/@apollo/protobufjs/src/index-minimal.node

Require stack:
- /var/app/current/.yarn/unplugged/@apollo-protobufjs-npm-1.0.5-a95d547915/node_modules/@apollo/protobufjs/minimal.js
- /var/app/current/.yarn/cache/apollo-reporting-protobuf-npm-0.6.2-87872b385d-6e00d36475.zip/node_modules/apollo-reporting-protobuf/dist/protobuf.js
- /var/app/current/.yarn/cache/apollo-reporting-protobuf-npm-0.6.2-87872b385d-6e00d36475.zip/node_modules/apollo-reporting-protobuf/dist/index.js
- /var/app/current/.yarn/$$virtual/apollo-server-core-virtual-c315e24bf7/0/cache/apollo-server-core-npm-2.22.2-56e07813ba-a2a3d0189e.zip/node_modules/apollo-server-core/dist/plugin/usageReporting/plugin.js
- /var/app/current/.yarn/$$virtual/apollo-server-core-virtual-c315e24bf7/0/cache/apollo-server-core-npm-2.22.2-56e07813ba-a2a3d0189e.zip/node_modules/apollo-server-core/dist/plugin/usageReporting/index.js
- /var/app/current/.yarn/$$virtual/apollo-server-core-virtual-c315e24bf7/0/cache/apollo-server-core-npm-2.22.2-56e07813ba-a2a3d0189e.zip/node_modules/apollo-server-core/dist/plugin/index.js
- /var/app/current/.yarn/$$virtual/apollo-server-core-virtual-c315e24bf7/0/cache/apollo-server-core-npm-2.22.2-56e07813ba-a2a3d0189e.zip/node_modules/apollo-server-core/dist/ApolloServer.js
- /var/app/current/.yarn/$$virtual/apollo-server-core-virtual-c315e24bf7/0/cache/apollo-server-core-npm-2.22.2-56e07813ba-a2a3d0189e.zip/node_modules/apollo-server-core/dist/index.js
- /var/app/current/.yarn/$$virtual/apollo-server-express-virtual-0b17a3c730/0/cache/apollo-server-express-npm-2.22.2-6833b00a46-c54a3e792a.zip/node_modules/apollo-server-express/dist/index.js
- /var/app/current/lib/server/index.js
- /var/app/current/lib/scripts/www.js
    at internalTools_makeError (/var/app/current/.pnp.js:36256:34)
    at resolveUnqualified (/var/app/current/.pnp.js:37295:13)
    at resolveRequest (/var/app/current/.pnp.js:37319:14)
    at Object.resolveRequest (/var/app/current/.pnp.js:37391:26)
    at Function.external_module_.Module._resolveFilename (/var/app/current/.pnp.js:36489:34)
    at Function.external_module_.Module._load (/var/app/current/.pnp.js:36354:48)
    at Module.require (internal/modules/cjs/loader.js:1044:19)
    at require (internal/modules/cjs/helpers.js:77:18)
    at Object.<anonymous> (/var/app/current/.yarn/unplugged/@apollo-protobufjs-npm-1.0.5-a95d547915/node_modules/@apollo/protobufjs/minimal.js:4:18)
    at Module._compile (internal/modules/cjs/loader.js:1158:30)
/var/app/current/.pnp.js:36512
    throw firstError;
```

**Comment by lensbart** [comment]
@instantlinux if the issue only arises in CI, check if your ignore file (e.g. `.ebignore` for AWS Elastic Beanstalk) matches the path not being found.

In my case, I had `src` in my ignore file instead of `./src`, so it left out some subdirectory from a subdependency.

I arrived here via Google, so maybe it’s not related to this issue, but it cost me a few hours to figure out, so I figured I’d share, just in case.
